sql: improve interface for injecting descriptors into internal executor

thoszhang opened the pull request:
As part of validating new schema elements (unique indexes, constraints)
in the schema changer, we inject in-memory descriptors to be used by the
internal executor that are never written to disk.

This is currently done by creating an entirely new dummy
`descs.Collection` and adding the injected descriptors as uncommitted
descriptors, and then setting this collection as the `tcModifier` on the
internal executor.  Then all subsequent queries using the internal
executor, which each get their own child `connExecutor` and
`descs.Collection`, will have their collection's uncommitted descriptors
replaced by the ones in the dummy collection.

This commit introduces the concept of a "synthetic descriptor" to refer
to these injected descriptors, and slightly improves the interface in
two ways:

1. Instead of creating a new `descs.Collection` to hold synthetic
   descriptors to copy, we now just use a slice of
   `catalog.Descriptor`s.
2. Synthetic descriptors are now made explicit in the `descs.Collection`
   and precede uncommitted descriptors when resolving immutable
   descriptors. Resolving these descriptors mutably is now illegal and
   will return an assertion error.

This commit doesn't change the fact that the synthetic descriptors to be
injected into each query/statement's child `descs.Collection` are set
on the internal executor itself. This is still not threadsafe. It would
make more sense for these descriptors to be scoped at the level of
individual queries.

Related to #34304.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58906)
<!-- Reviewable:end -->

Commit:
sql: improve interface for injecting descriptors into internal executor

As part of validating new schema elements (unique indexes, constraints)
in the schema changer, we inject in-memory descriptors to be used by the
internal executor that are never written to disk.

This is currently done by creating an entirely new dummy
`descs.Collection` and adding the injected descriptors as uncommitted
descriptors, and then setting this collection as the `tcModifier` on the
internal executor.  Then all subsequent queries using the internal
executor, which each get their own child `connExecutor` and
`descs.Collection`, will have their collection's uncommitted descriptors
replaced by the ones in the dummy collection.

This commit introduces the concept of a "synthetic descriptor" to refer
to these injected descriptors, and slightly improves the interface in
two ways:

1. Instead of creating a new `descs.Collection` to hold synthetic
   descriptors to copy, we now just use a slice of
   `catalog.Descriptor`s.
2. Synthetic descriptors are now made explicit in the `descs.Collection`
   and precede uncommitted descriptors when resolving immutable
   descriptors. Resolving these descriptors mutably is now illegal and
   will return an assertion error.

This commit doesn't change the fact that the synthetic descriptors to be
injected into each query/statement's child `descs.Collection` are set
on the internal executor itself. This is still not threadsafe. It would
make more sense for these descriptors to be scoped at the level of
individual queries.

Release note: None

lucy-zhang commented:
This could maybe use a few tests, but it's ready for a look. This interface could certainly be improved further, but I figure we'll need to expand the synthetic descriptor concept in new ways for transactional schema changes, and further movement in that direction now would be premature.

ajwerner reviewed (commented):
This is definitely better. 

> Reviewed 3 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58906#-:-MQyjH325x6ABSbHXWYE:b3svdgs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/catalog/descs/collection.go, line 361 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58906#-MQykT-aAHIeFAGnmQ9l:-MQykT-aAHIeFAGnmQ9m:b61uik2) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fb20bc9aec67711bbdf6e68cafe6ee94a81168/pkg/sql/catalog/descs/collection.go#L361)):*
> ```Go
> 
> 	getDatabaseByName := func() (found bool, _ catalog.Descriptor, err error) {
> 		if found, sd := tc.getSyntheticDescriptorByName(
> ```

could we pull the logic to retreive the synthetic descriptor underneath the logic to get uncommitted descriptors? 


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
Added a basic test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58906#-:-MQzXz0j8ynYyXMcSQhR:bvle2me)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/catalog/descs/collection.go, line 361 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58906#-MQykT-aAHIeFAGnmQ9l:-MQzXaZf5kXC3KpmR6ez:b-go2vbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fb20bc9aec67711bbdf6e68cafe6ee94a81168/pkg/sql/catalog/descs/collection.go#L361)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

could we pull the logic to retreive the synthetic descriptor underneath the logic to get uncommitted descriptors? 
</blockquote></details>

I did this for the resolution-by-name path. I think I'll do it for the by-ID path as part of #57343 once this merges.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58906)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/catalog/descs/collection.go, line 361 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58906#-MQykT-aAHIeFAGnmQ9l:-MR3Nhsg2RW7RvpZVeZs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fb20bc9aec67711bbdf6e68cafe6ee94a81168/pkg/sql/catalog/descs/collection.go#L361)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I did this for the resolution-by-name path. I think I'll do it for the by-ID path as part of #57343 once this merges.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
I rebased on top of some recent changes. This is RFAL.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r2, 1 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58906#-:-MSU8P0v0mhqGAbDAtqu:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

lucy-zhang commented:
TFTR

bors r+
